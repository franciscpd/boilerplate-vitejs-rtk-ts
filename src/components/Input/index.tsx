import { useMemo } from 'react'
import clx from 'classnames'
import {
  Form as FormControl,
  Input as InputBase,
  InputProps as InputBaseProps,
} from 'react-daisyui'

import Label from '../Label'

export type InputProps = {
  label?: string
  name: string
  error?: string | undefined
  fullWidth?: boolean | undefined
  labelClassName?: string
  errorClassName?: string
  formControlClassName?: string
} & InputBaseProps

function Input({
  label,
  fullWidth,
  name,
  error,
  className,
  labelClassName,
  errorClassName,
  formControlClassName,
  color: colorParam = 'primary',
  ...props
}: InputProps) {
  const isError = useMemo(() => !!error, [error])
  const color = useMemo(
    () => (isError ? 'error' : colorParam),
    [colorParam, isError]
  )

  return (
    <FormControl
      className={clx('mt-4', { 'w-full': fullWidth }, formControlClassName)}
    >
      {label && (
        <Label error={isError} htmlFor={name} className={labelClassName}>
          {label}
        </Label>
      )}
      <InputBase
        color={color}
        name={name}
        {...props}
        className={clx(
          'text-gray-500',
          className,
          `focus:outline-offset-0 focus:outline-0 focus:border-${color} focus:ring-${color} focus:shadow-none`
        )}
      />
      {error && (
        <Label error={isError} alt className={errorClassName}>
          {error}
        </Label>
      )}
    </FormControl>
  )
}

export default Input
