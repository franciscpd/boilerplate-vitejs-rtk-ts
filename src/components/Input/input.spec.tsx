import React from 'react'
import { describe, test, expect } from 'vitest'

import { render } from '../../tests/utils'

import Input from '.'

describe('Input', () => {
  test('Should render the input', () => {
    const { getByPlaceholderText, container } = render(
      <Input name="email" label="E-mail" placeholder="email@test.com" />
    )

    expect(getByPlaceholderText('email@test.com')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  test('Should render color primary correct', () => {
    const { container } = render(
      <Input name="email" label="E-mail" placeholder="email@test.com" />
    )

    expect(container.querySelector('.label-text')).toHaveClass('text-primary')
    expect(container.querySelector('.input')).toHaveClass('input-primary')
    expect(container.querySelector('.input')).toHaveClass(
      'focus:border-primary'
    )
    expect(container.querySelector('.input')).toHaveClass('focus:ring-primary')
  })

  test('Should render color error correct', () => {
    const { container } = render(
      <Input
        name="email"
        label="E-mail"
        placeholder="email@test.com"
        error="Field required"
      />
    )

    expect(container.querySelector('.label-text')).toHaveClass('text-error')
    expect(container.querySelector('.input')).toHaveClass('input-error')
    expect(container.querySelector('.input')).toHaveClass('focus:border-error')
    expect(container.querySelector('.input')).toHaveClass('focus:ring-error')
    expect(container.querySelector('.label-text-alt')).toHaveClass('text-error')
  })

  test('Should render custom color correct', () => {
    const { container } = render(
      <Input
        name="email"
        label="E-mail"
        placeholder="email@test.com"
        color="info"
        labelClassName="text-info"
      />
    )

    expect(container.querySelector('.label-text')).toHaveClass('text-info')
    expect(container.querySelector('.input')).toHaveClass('input-info')
    expect(container.querySelector('.input')).toHaveClass('focus:border-info')
    expect(container.querySelector('.input')).toHaveClass('focus:ring-info')
  })
})
