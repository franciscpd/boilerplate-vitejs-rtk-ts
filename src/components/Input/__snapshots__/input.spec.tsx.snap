// Vitest Snapshot v1

exports[`Input > Should render the input 1`] = `
<div
  class="form-control mt-4"
  role="form"
>
  <label
    class="label"
    for="email"
  >
    <span
      class="label-text text-primary"
    >
      E-mail
    </span>
  </label>
  <input
    class="input text-gray-500 focus:outline-0 focus:border-primary focus:ring-primary focus:shadow-none input-primary focus:outline-offset-0 input-bordered"
    name="email"
    placeholder="email@test.com"
    value=""
  />
</div>
`;
